Close XPC connection between UI process and WebContent process

pvollan opened the pull request:
#### fbbc247c4f4f9582a6d2bda8cc22f7679035062e
<pre>
Close XPC connection between UI process and WebContent process
<a href="https://bugs.webkit.org/show_bug.cgi?id=269574">https://bugs.webkit.org/show_bug.cgi?id=269574</a>
<a href="https://rdar.apple.com/115840818">rdar://115840818</a>

Reviewed by Chris Dumez.

Close XPC connection between UI process and WebContent process after launch, since the XPC connection
should no longer be needed then. This feature flag can only be enabled when WebKit processes are
launched as extensions, since the lifetime of XPC services are closely related to the lifetime of the
XPC connection. The audit token needs to be cached in the Connection object, since it cannot be
retrieved after the XPC connection has been closed. This patch also adds a missing call to the
completion handler in maybeInitializeSandboxExtensionHandle, which was observed during the work on
this patch. This change has been validated locally on iOS where WebKit processes are launched as
extensions.

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebKit/Platform/IPC/Connection.h:
* Source/WebKit/Platform/IPC/cocoa/ConnectionCocoa.mm:
(IPC::Connection::platformOpen):
(IPC::Connection::getAuditToken):
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceMain.mm:
(WebKit::XPCServiceEventHandler):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::maybeInitializeSandboxExtensionHandle):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::didClose):
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::platformInitializeWebProcess):

Canonical link: <a href="https://commits.webkit.org/284270@main">https://commits.webkit.org/284270@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6f42cc269032f61408e8f2bfb0a02b2819512503

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68797 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48191 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21459 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/72868 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19943 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55988 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19762 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54835 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13271 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71863 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44029 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59406 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35304 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40694 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16833 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18305 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/61914 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62644 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17180 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74563 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/68044 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12771 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16428 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62323 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12810 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59488 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62359 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10320 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3931 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/89823 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10520 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43989 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15911 "Found 101 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Basics/DefProp.js.default, ChakraCore.yaml/ChakraCore/test/es6/ES6Function_bugs.js.default, ChakraCore.yaml/ChakraCore/test/es6/superDotOSBug3930962.js.default, ChakraCore.yaml/ChakraCore/test/fieldopts/fixedfieldmonocheck6.js.default, ChakraCore.yaml/ChakraCore/test/strict/16.catch_sm.js.default, jsc-layout-tests.yaml/js/script-tests/dfg-arguments-alias-one-block.js.layout, jsc-layout-tests.yaml/js/script-tests/dfg-double-use-of-post-simplification-double-prediction.js.layout, stress/arguments-define-property.js.dfg-eager, stress/arguments-non-configurable.js.dfg-eager, stress/array-species-functions.js.bytecode-cache ... (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45066 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46261 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44805 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez reviewed (commented):
Are we sure this connection is not needed? If I look at the termination watchdog for e.g.:
```
void XPCConnectionTerminationWatchdog::watchdogTimerFired()
{
    terminateWithReason(m_xpcConnection.get(), ReasonCode::WatchdogTimerFired, "XPCConnectionTerminationWatchdog::watchdogTimerFired");
    delete this;
}
```

Looks like we need an XPC connection to terminate the child process? Will this still work if the XPC connection has been closed?

cdumez commented on Source/WebKit/Platform/cocoa/XPCUtilities.mm:
```diff
@@ -39,4 +41,9 @@ void terminateWithReason(xpc_connection_t connection, ReasonCode, const char*)
 
+void terminateWithReason(pid_t pid, ReasonCode reasonCode, const char* reason)
+{
+    terminate_with_reason(pid, OS_REASON_WEBKIT, static_cast<uint64_t>(reasonCode), reason, OS_REASON_FLAG_NO_CRASH_REPORT);
```
Are you certain this really works? I see to remember Alex trying this in the past and it not working because the system wouldn't let us kill other processes by PID, even child processes.

pvollan replied:
Ah, good point, I think the blocker has been resolved, looking closer.

Thanks for reviewing!

cdumez replied:
I hope you validated locally that this is actually working on both macOS and iOS.

pvollan replied:
I have only validated this on iOS, since we are still using XPC services on macOS. In the XPC case, this behavior cannot be enabled, I believe.

Thanks for reviewing!

cdumez commented on Source/WebKit/UIProcess/Cocoa/XPCConnectionTerminationWatchdog.mm:
```diff
@@ -38,7 +38,11 @@
     
 XPCConnectionTerminationWatchdog::XPCConnectionTerminationWatchdog(AuxiliaryProcessProxy& process, Seconds interval)
+#if ENABLE(CLOSE_XPC_CONNECTION_POST_LAUNCH)
+    : m_processID(process.processID())
```
The previous code was working even if the process was still launching. The new code doesn't. Sounds like this may be an issue.

pvollan replied:
Ah, good catch. I will try to address this.

cdumez replied:
I don't think this issue was addressed in the latest iteration?

pvollan replied:
Ah, good point! I believe this issue has been addressed in the latest patch, where an audit token is used instead to avoid PID races.

cdumez commented on Source/WebKit/UIProcess/Cocoa/XPCConnectionTerminationWatchdog.mm:
```diff
@@ -47,7 +52,12 @@
+    auto returnValue = proc_signal_with_audittoken(&m_auditToken, SIGKILL);
+    RELEASE_LOG(Process, "XPCConnectionTerminationWatchdog::watchdogTimerFired(): proc_signal_with_audittoken returned %d", returnValue);
+#else
     terminateWithReason(m_xpcConnection.get(), ReasonCode::WatchdogTimerFired, "XPCConnectionTerminationWatchdog::watchdogTimerFired");
```
It is very suspect for XPCConnectionTerminationWatchdog to be using another API to terminate processes than other code paths. Other code paths still rely on `terminateWithReason()`, which AFAICT will no longer work once `CLOSE_WEBCONTENT_XPC_CONNECTION_POST_LAUNCH` gets turned on.

pvollan replied:
Ah, great catch, Chris! I will look closer at this.

Thanks for reviewing!

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -1226,6 +1226,10 @@ bool WebProcessProxy::dispatchSyncMessage(IPC::Connection& connection, IPC::Deco
 
 void WebProcessProxy::didClose(IPC::Connection& connection)
 {
+#if ENABLE(CLOSE_WEBCONTENT_XPC_CONNECTION_POST_LAUNCH)
```
I'm a little confused why we need this and I'm worried this may cause us to ignore certain crashes.
This is supposed to get called when the IPC connection gets closed. Our IPC connection uses mach and not xpc, therefore, why does it get called if we kill the xpc_connection but do not terminate the child process? Can't we avoid this getting called at all in this case?

What if this is a real child process crash that happened on launch and we haven't closed the xpc connection yet?

pvollan replied:
Good point! It turns out this method is not called with the current location of the `xpc_connection_cancel` call, so I have removed it in the latest version.

Thanks for reviewing!

cdumez commented on Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceMain.mm:
```diff
@@ -135,9 +137,16 @@ void XPCServiceEventHandler(xpc_connection_t peer)
-                    }];
+                    bool shouldExit = true;
+#if ENABLE(CLOSE_WEBCONTENT_XPC_CONNECTION_POST_LAUNCH)
+                    if (s_isWebProcess)
```
Maybe we can do an early return and avoid the `shouldExit` local variable?
```
#if ENABLE(CLOSE_WEBCONTENT_XPC_CONNECTION_POST_LAUNCH)
if (s_isWebProcess)
    return;
#endif
```

cdumez commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -577,6 +577,11 @@ static void setVideoDecoderBehaviors(OptionSet<VideoDecoderBehavior> videoDecode
 #endif
+
+#if ENABLE(CLOSE_WEBCONTENT_XPC_CONNECTION_POST_LAUNCH)
+    auto connection = parentProcessConnection()->xpcConnection();
```
I don't think we need a local variable here.

pvollan committed:
Close XPC connection between UI process and WebContent process
https://bugs.webkit.org/show_bug.cgi?id=269574
rdar://115840818

Reviewed by Chris Dumez.

Close XPC connection between UI process and WebContent process after launch, since the XPC connection
should no longer be needed then. This feature flag can only be enabled when WebKit processes are
launched as extensions, since the lifetime of XPC services are closely related to the lifetime of the
XPC connection. The audit token needs to be cached in the Connection object, since it cannot be
retrieved after the XPC connection has been closed. This patch also adds a missing call to the
completion handler in maybeInitializeSandboxExtensionHandle, which was observed during the work on
this patch. This change has been validated locally on iOS where WebKit processes are launched as
extensions.

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebKit/Platform/IPC/Connection.h:
* Source/WebKit/Platform/IPC/cocoa/ConnectionCocoa.mm:
(IPC::Connection::platformOpen):
(IPC::Connection::getAuditToken):
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceMain.mm:
(WebKit::XPCServiceEventHandler):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::maybeInitializeSandboxExtensionHandle):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::didClose):
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::platformInitializeWebProcess):

Canonical link: https://commits.webkit.org/284270@main
